File: Algorithms/T208_Implement_Trie.java
class Trie {
    Trie[] children = new Trie[26];
    boolean end = false;
    
    /** Initialize your data structure here. */
    public Trie() {
        
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        Trie cur = this;        
        for(int i = 0; i < word.length(); i++){
            if(cur.children[word.charAt(i) - 'a'] != null){
                cur = cur.children[word.charAt(i) - 'a'];
            }else{
                Trie next = new Trie();
                cur.children[word.charAt(i) - 'a'] = next;
                cur = next;
            }               
        } 
        cur.end = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        Trie cur = this; 
        for(int i = 0; i < word.length(); i++){
            Trie next = cur.children[word.charAt(i) - 'a'];
            if(next == null) return false;
            cur = next;
        }
        return cur.end;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        Trie cur = this; 
        for(int i = 0; i < prefix.length(); i++){
            Trie next = cur.children[prefix.charAt(i) - 'a'];
            if(next == null) return false;
            cur = next;
        }
        return true;
    }
}

File: Algorithms/C++/T208_Implement_Trie.cpp
class Trie {
public:
    vector<vector<int>> v;
    unordered_set<int> s;
    /** Initialize your data structure here. */
    Trie() {
        v.push_back(vector<int>(26, -1));
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        int pos = 0;
        for(auto c : word){
            if(v[pos][c - 'a'] != -1){
                pos = v[pos][c - 'a'];
            }else{
                v.push_back(vector<int>(26, -1));
                v[pos][c - 'a'] = v.size() - 1;
                pos = v.size() - 1;
            }
        }
        s.insert(pos);
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        int pos = 0;
        for(auto c : word){
            if(v[pos][c - 'a'] != -1){
                pos = v[pos][c - 'a'];
            }else{
                return false;
            }

